src/lua/alloc.c

```diff
@@ -0,0 +1,171 @@
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "lua/utils.h"
+
+#include <lauxlib.h> /* luaL_error */
+
+#define LUA_MEMORY_DEFAULT (2LL * 1024 * 1024 * 1024)
+
+/**
+ * The default allocator function which is wrapped into a new one
+ * with the Lua memory limit checker.
+ */
+static lua_Alloc orig_alloc;
+/**
+ * The maximum allowed LuaJIT memory limit.
+ */
+static size_t memory = LUA_MEMORY_DEFAULT;
+/**
+ * Currently allocated memory amount by LuaJIT.
+ */
+static size_t allocated;
+
+/**
+ * Lua custom memory allocation function. It extends the original
+ * one with a memory counter and a limit check.
+ */
+static void *
+luaT_alloc(void *ud, void *ptr, size_t osize, size_t nsize)
+{
+	size_t new_allocated = allocated + nsize - osize;
+	if (new_allocated > memory) {
+		diag_set(OutOfMemory, nsize, "luaT_alloc", "Lua objects");
+		return NULL;
+	}
+
+	void *result = orig_alloc(ud, ptr, osize, nsize);
+
+	if (result != NULL || nsize == 0)
+		allocated = new_allocated;
+	else
+		diag_set(OutOfMemory, nsize, "luaT_alloc", "Lua objects");
+
+	return result;
+}
+
+int
+luaT_setallocmemory(size_t amount)
+{
+	if (amount < allocated) {
+		diag_set(LuajitError, "Cannot limit the Lua memory with values "
+			 "less than the currently allocated amount");
+		return -1;
+	}
+	memory = amount;
+	return 0;
+}
+
+void
+luaT_initalloc(struct lua_State *L)
+{
+	assert(L != NULL);
+
+	allocated = luaL_getgctotal(L);
+
+	void *orig_ud;
+	orig_alloc = lua_getallocf(L, &orig_ud);
+
+	assert(orig_alloc != NULL);
+
+	lua_setallocf(L, luaT_alloc, orig_ud);
+}
+
+/**
+ * alloc.getmemory() -- get the allocator memory limit.
+ */
+static int
+lbox_alloc_getmemory(struct lua_State *L)
+{
+	lua_pushinteger(L, memory);
```

Comment:
Should we use `luaL_pushuint64` or something like it? The `memory` variable may go beyond 2^31-1.

Reply:
Seems like the following methods should be ok, I've stolen the approach from 
```lua
static int
luaT_httpc_io_read(lua_State *L)
{
	struct httpc_io *io = luaT_httpc_checkio(L);
	struct httpc_request *req = io->req;
	uint32_t ctypeid = 0;
	char *buf = *(char **)luaL_checkcdata(L, 2, &ctypeid);
	ssize_t len = luaL_checkinteger(L, 3);
	double timeout = luaL_checknumber(L, 4);

	if (len <= 0) {
		diag_set(IllegalParams, "io: payload length must be >= 0");
		return luaT_error(L);
	}

	if (timeout < 0) {
		diag_set(IllegalParams, "io: timeout must be >= 0");
		return luaT_error(L);
	}

	ssize_t res = httpc_request_io_read(req, buf, len, timeout);
	if (res < 0) {
		return luaT_error(L);
	}

	lua_pushinteger(L, res);
	return 1;
}
```
AFAIU, Lua uses 64-bit integers by default (though it can be compiled in 32-bit mode). I suppose there is no need to take it into account since either Lua is compiled in GC64 mode and the option is works OK or Lua memory limit can't be more than 2^31 - 1.

Reply:
Lua has no integers, just `number` (which is `double` under the hood).

`lua_pushinteger` accepts `lua_Integer`, which is `LUA_INTEGER` defined as `ptrdiff_t`. Hm. Seems like it should work with integers beyond 2^32-1.

OTOH, it likely has rounding problems for values above 2^53. It is ~8PiB. Hm. Maybe we can pass it over.

I would like to hear @Buristan's opinion.